change-notes/1.23/analysis-cpp.md

```diff
@@ -2,44 +2,40 @@
-| Too many arguments to formatting function (`cpp/too-many-format-arguments`) | Fewer false positive results | Fixed false positives resulting from mistmatching declarations of a formatting function. |
-| Unclear comparison precedence (`cpp/comparison-precedence`) | Fewer false positive results | False positives involving template classes and functions have been fixed. |
-| Comparison of narrow type with wide type in loop condition (`cpp/comparison-with-wider-type`) | Higher precision | The precision of this query has been increased to "high" as the alerts from this query have proved to be valuable on real-world projects. With this precision, results are now displayed by default in LGTM. |
-| Non-constant format string (`cpp/non-constant-format`) | Fewer false positive results | Fixed false positives resulting from mistmatching declarations of a formatting function. |
-| Wrong type of arguments to formatting function (`cpp/wrong-type-format-argument`) | More correct results and fewer false positive results | This query now understands explicitly specified argument numbers in format strings, such as the `1$` in `%1$s`. |
+| Too few arguments to formatting function (`cpp/wrong-number-format-arguments`) | Fewer false positive results | Fixed false positive results triggered by mismatching declarations of a formatting function. |
+| Too many arguments to formatting function (`cpp/too-many-format-arguments`) | Fewer false positive results | Fixed false positive results triggered by mismatching declarations of a formatting function. |
+| Unclear comparison precedence (`cpp/comparison-precedence`) | Fewer false positive results | False positive results involving template classes and functions have been fixed. |
+| Wrong type of arguments to formatting function (`cpp/wrong-type-format-argument`) | More correct results and fewer false positive results | This query now understands explicitly-specified argument numbers in format strings, such as the `1$` in `%1$s`. |
 
 ## Changes to libraries
 
 * The data-flow library in `semmle.code.cpp.dataflow.DataFlow` and
   `semmle.code.cpp.dataflow.TaintTracking` have had extensive changes:
   * Data flow through fields is now more complete and reliable.
-  * To debug a lack of flow, instead of specifying `isSink(Node n) { any() }`
-    on a configuration to
-    explore the possible flow from a source, it is recommended to use the new
-    `Configuration::hasPartialFlow` predicate as this gives a more complete
-    picture of the partial flow paths from a given source. The feature is
-    disabled by default and can be enabled for individual configurations by
-    overriding `int explorationLimit()`.
+  * The data-flow library has been extended with a new feature to aid debugging. 
+    If you want to explore the possible flow from a source, replace
+    `isSink(Node n) { any() }` with the new `Configuration::hasPartialFlow` predicate. 
```

Comment:
In the second sentence, `replace` sounds like it should be taken literally, but it shouldn't. Writing `isSink(Node n) { any() }` was previously a good way to debug data flow, but not everyone did that, so there will be nothing to replace in many cases. Instead, a call to `hasPartialFlow` should literally replace the call to `hasFlow` or `hasFlowPath`.

I propose this text instead for the second sentence:

To explore the possible flow from all sources, including flow that doesn't reach any sink, use the `hasPartialFlow` predicate on `Configuration` instead of `hasFlow` or `hasFlowPath`.

@aschackmull and @hvitved, what do you think?

Reply:
I suggest using the original change not from C#:  https://github.com/Semmle/ql/pull/2430/files#diff-95682413e8e7d29056c3f97a3d0b8f43L38-L40

Reply:
I agree. The original C# change note was clear.

Reply:
I was trying to reorganize the sentence so that it started with what the user was trying to do, rather than how they'd coded it. As well as making the sentence shorter. Clearly my changes were bad. Would this work any better?

Previously, to explore the possible flow from all sources you could specify `isSink(Node n) { any() }` on a configuration. Now you can use the new `Configuration::hasPartialFlow` predicate, which gives a more complete picture of the partial flow paths from a given source, including flow that doesn't reach any sink.

Reply:
LGTM

Reply:
@felicitymay Your latest proposal sounds good to me. I assume it will still be preceded by "The data-flow library has been extended with a new feature to aid debugging." and followed by "The feature is disabled ...".

Reply:
Thanks for everyone's patience with this. Now that I'm out of meetings, I'll update this PR and the other change notes.